### PR TITLE
fix geometry validator

### PR DESCRIPTION
**Pull request by vcloarec**
fix QgsGeometryVaidatot that can make Qgis crash when manipulating vertices

### CODE REVIEW

**Comment by nyalldawson** [comment]
Ouch -- annoying that this could compile :( 

Can you add a test which triggers this? QgsGeometryValidator unfortunately has very poor test coverage at the moment

**Comment by vcloarec** [comment]
I found this one during own GIS work, and I do not have much time for now to add test...

File: src/core/qgsgeometryvalidator.cpp [hunk 1]
```diff
@@ -65,8 +65,8 @@ void QgsGeometryValidator::checkRingIntersections( int p0, int i0, const QgsLine
         {
           d = -distLine2Point( ring1XAtj, ring1YAtj, w.perpVector(), sX, sY );
           if ( d > 0 && d < w.length() &&
-               ring0[i + 1] != ring1[j + 1] && ring0[i + 1] != ring1[j] &&
-               ring0[i + 0] != ring1[j + 1] && ring0[i + 0] != ring1[j] )
+               ring0->pointN( i + 1 ) != ring1->pointN( j + 1 ) && ring1->pointN( i + 1 ) != ring1->pointN( j ) &&
```
Comment (uclaros) [inline src/core/qgsgeometryvalidator.cpp:68]: Seems like you've mixed the rings. Ring0's counter is i and ring1's counter is j

Comment (vcloarec) [reply src/core/qgsgeometryvalidator.cpp:68]: oups... you are right, it is a mess!

Comment (vcloarec) [reply src/core/qgsgeometryvalidator.cpp]: fixed

File: src/core/qgsgeometryvalidator.cpp [hunk 2]
```diff
@@ -65,8 +65,8 @@ void QgsGeometryValidator::checkRingIntersections( int p0, int i0, const QgsLine
         {
           d = -distLine2Point( ring1XAtj, ring1YAtj, w.perpVector(), sX, sY );
           if ( d > 0 && d < w.length() &&
-               ring0[i + 1] != ring1[j + 1] && ring0[i + 1] != ring1[j] &&
-               ring0[i + 0] != ring1[j + 1] && ring0[i + 0] != ring1[j] )
+               ring0->pointN( i + 1 ) != ring1->pointN( j + 1 ) && ring0->pointN( i + 1 ) != ring1->pointN( j ) &&
+               ring0->pointN( i + 0 ) != ring1->pointN( j + 1 ) && ring0->pointN( i + 0 ) != ring1->pointN( j ) )
           {
             QString msg = QObject::tr( "segment %1 of ring %2 of polygon %3 intersects segment %4 of ring %5 of polygon %6 at %7, %8" )
                           .arg( i0 ).arg( i ).arg( p0 )
```
Comment (uclaros) [inline src/core/qgsgeometryvalidator.cpp:72]: Seems like those args are mixed too. i is the vertex index and i0 should be the ring index

**Comment by nyalldawson** [comment]
I've added a test in https://github.com/qgis/QGIS/pull/39790 (doing my penance)

**Pull request closed without merging**